### PR TITLE
Stein Variational Gradient Descent

### PR DESCRIPTION
**Pull request by twiecki**
Based on: https://github.com/DartML/Stein-Variational-Gradient-Descent
Paper: https://arxiv.org/abs/1608.04471

Still running some experiments but seems to work quite well.

### CODE REVIEW

**Comment by springcoil** [comment]
Looks good to me so far. I can't wait to dig into the implementation :)

**Comment by twiecki** [comment]
Small example using @fonnesbeck's GP: https://github.com/pymc-devs/pymc3/blob/1a0c66b4938f79d685b8df953db49808918e976f/docs/source/notebooks/gaussian_process.ipynb

**Comment by taku-y** [comment]
Very good example! I will take a closer look at the code tonight.
Do you have a plan to support mini-batch?

**Comment by fonnesbeck** [comment]
Very cool. I wonder if we can parallelize the gradient calculation over the particles? I would think that would speed things up considerably.

File: pymc3/variational/svgd.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    if start is None:
+        start = model.test_point
+    start = model.dict_to_array(start)
+
+    # Initialize particles
+    x0 = np.tile(start, (n_particles, 1))
+    x0 += np.random.normal(0, jitter, x0.shape)
+
+    theta = tt.dmatrix('theta')
+    theta.tag.test_value = x0
```
Comment (fonnesbeck) [inline pymc3/variational/svgd.py]: Isn't this where Shared variables should be used instead of tensors (like you would the weights of a neural network)? Seems like theta is an internal state.

File: pymc3/variational/svgd.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    for i in progress:
+        grad_theta = svgd_grad(theta)[0]
+
+        # adagrad
+        if iter == 0:
+            historical_grad = historical_grad + grad_theta ** 2
+        else:
+            historical_grad = alpha * historical_grad + (1 - alpha) * (grad_theta ** 2)
+        adj_grad = np.divide(grad_theta, fudge_factor+np.sqrt(historical_grad))
+        theta = theta + stepsize * adj_grad
```
Comment (fonnesbeck) [inline pymc3/variational/svgd.py]: Shouldn't the stepsize get smaller as the procedure iterates, or does it not matter?

**Comment by twiecki** [comment]
@taku-y Yes, I'd like to support mini-batch, should not be too difficult but haven't looked at advi_minibatch enough. Is the local vs global RV distinction important here? How much of the API do you think we could reuse?

File: pymc3/variational/svgd.py (hunk 1 above)
Comment (twiecki) [reply pymc3/variational/svgd.py]: Yes, good point.

**Comment by taku-y** [comment]
@twiecki The distinction between the local and global RVs is related to autoencoding VB, not to mini-batch. Since I don't understand the SVGD code, I'm not sure reusing the advi_minibatch code is good or not good, sorry.

**Comment by twiecki** [comment]
@fonnesbeck Turned `theta` into a shared, which allowed some other simplifications.

> Very cool. I wonder if we can parallelize the gradient calculation over the particles? I would think that would speed things up considerably.

I think it's already doing that, at least it's using all my cores when running it.

**Comment by twiecki** [comment]
OK, thanks to @DartML we now have ADAM as an optimizer here.
Will add tests and docs and I think then we can merge.

**Comment by ferrine** [comment]
I'm finishing #1600. What about further refactoring SVGD with this API?

**Comment by twiecki** [comment]
Yes, we definitely should. Any pointers?

**Comment by ferrine** [comment]
It seems that optimizers cant, take list of shared for computing updates, could you refactor that please? Lasagne updates are good example of such implementation

**Comment by twiecki** [comment]
Could you be more specific as to how you'd want to pass those in?

**Comment by ferrine** [comment]
I mean this API ([link](https://github.com/Lasagne/Lasagne/blob/master/lasagne/updates.py#L545)) as it is nat always practical to have only one shared variable
```python
# source: lasagne.updates
def adam(loss_or_grads, params, learning_rate=0.001, beta1=0.9,
         beta2=0.999, epsilon=1e-8):
    """
    Parameters
    ----------
    loss_or_grads : symbolic expression or list of expressions
        A scalar loss expression, or a list of gradient expressions
    params : list of shared variables
        The variables to generate update expressions for
    ...
    """
```

**Comment by twiecki** [comment]
This PR is currently broken (as of merging #1612), it does not converge to the correct posterior anymore. Perhaps @DartML has an idea.

**Comment by twiecki** [comment]
Closing in favor of https://github.com/pymc-devs/pymc3/pull/1671.
